kvm system ioctls: updated public functions docs

andreeaflorescu opened the pull request:
## Changes
Added examples and more details for public functions.
Also renamed get_run_size() to run_size().

## Note:
`cargo test` will also run the documentation examples. Even if some examples are simple and just call the function it can help us in the future to keep the documentation updated as we make code changes (cargo test fails otherwise).

## How to test it locally
The best way to review this PR is by generating the documentation as follows:

```
# Clone Repo
git clone git@github.com:rust-vmm/kvm-ioctls.git
cd kvm-ioctls
# Fetch the code from my fork 
git fetch https://github.com/andreeaflorescu/kvm-ioctls.git pub_docs
git checkout FETCH_HEAD
# Generate and open the documentation 
cargo doc --open
```

## Still missing

High level crate design & usage. Will add this as part of fixing #21

alxiord commented on src/ioctls/system.rs:
````diff
@@ -81,12 +114,23 @@ impl Kvm {
+    ///
+    /// ```
+    /// # use kvm_ioctls::Kvm;
+    /// use kvm_ioctls::Cap;
````
Why does `use Kvm` need `#` while `use Cap` doesn't?

andreeaflorescu replied:
use Cap is firstly used in the comments here. Future references to Cap will not include this line. Same happens fro use Kvm. But I can also comment both.

alxiord commented on src/ioctls/system.rs:
````diff
@@ -96,6 +140,14 @@ impl Kvm {
+    /// assert!(kvm.get_vcpu_mmap_size().unwrap() > 0);
+    /// ```
+    ///
     pub fn get_vcpu_mmap_size(&self) -> Result<usize> {
````
Since you got rid of the `get_` prefix from `get_run_size`, shouldn't the getter naming convention be consistent?

andreeaflorescu replied:
get_run_size is named that way because of the name of the ioctl: KVM_GET_VCPU_MMAP_SIZE. All ioctl methods are named by removing the KVM_ prefix and converting it to lower case.

alxiord commented on src/ioctls/system.rs:
```diff
@@ -187,6 +279,19 @@ impl Kvm {
+    ///     let mut cpuid = kvm.get_supported_cpuid(MAX_KVM_CPUID_ENTRIES).unwrap();
+    ///     let cpuid_entries = cpuid.mut_entries_slice();
+    ///     assert!(cpuid_entries.len() <= MAX_KVM_CPUID_ENTRIES);
+    ///
```
nit: remove blank line

andreeaflorescu replied:
Will do, waiting for more comments before doing a push.

sameo commented on src/ioctls/vm.rs:
```diff
@@ -247,7 +247,7 @@ impl VmFd {
 
     /// Get the `kvm_run` size.
-    pub fn get_run_size(&self) -> usize {
+    pub fn run_size(&self) -> usize {
```
This change should be a separate commit. Maybe even a separate PR.

andreeaflorescu replied:
I'll add it to a different commit (I was lazy).

The merge commit will vanish once we merge the PR to master. That's how we also do it in Firecracker to avoid `push -f` when the only thing you want to do is update the branch to master.

Trying to find an official documentation that explains this but totally failing.

sameo replied:
> avoid push -f when the only thing you want to do is update the branch to master.

I figured that much. I personally don't mind push forcing to a development branch in order to keep the reviewed commit history clean.

sameo reviewed (commented):
Please split the `run_size()` change into at least its own commit.
Also, one of the commit is a merge point, and we should avoid that.

dianpopa commented on src/cap.rs:
```diff
@@ -9,7 +9,11 @@ use kvm_bindings::*;
 
 /// Enumeration of the extension capability list that KVM exposes.
```
Since cargo doc will put these under "Enums" category, I would remove "Enumeration of" and I would leave only "The list of capabilities exposed by KVM" or something of the kind.

dianpopa commented on src/cap.rs:
```diff
@@ -9,7 +9,11 @@ use kvm_bindings::*;
+/// capability is available.
+///
+/// The list of capabilities is based on the the KVM_CAP_* defines from the
 /// [Linux KVM header](https://elixir.bootlin.com/linux/latest/source/include/uapi/linux/kvm.h).
```
Could we do the same thing (put a link to the elixir sources) for the `VpuExit` documentation. Currently it looks like this : `The exit reasons are mapped to the KVM_EXIT_* defines from include/uapi/linux/kvm.h.` and it looks kind of ugly when doing `cargo doc --open`.

dianpopa commented on src/ioctls/system.rs:
```diff
@@ -25,6 +25,13 @@ pub struct Kvm {
 impl Kvm {
     /// Opens `/dev/kvm/` and returns a `Kvm` object on success.
```
```suggestion
    /// Opens `/dev/kvm` and returns a `Kvm` object on success.
```

dianpopa commented on src/ioctls/system.rs:
```diff
@@ -48,10 +57,25 @@ impl Kvm {
     ///
+    /// One usecase for this method is opening `/dev/kvm` before exec-ing into a
+    /// process with seccomp filters enabled that blacklist the `sys_open` syscall.
+    /// For this usecase `open_with_cloexec` must be called with the`close_on_exec`
```
```suggestion
    /// For this usecase `open_with_cloexec` must be called with the `close_on_exec`
```

dianpopa reviewed (commented):
Additional comments related to documentation:
* `set_user_memory_region` does not have arguments section.
* `create_pit2` does not have arguments section

andreeaflorescu committed:
added documentation for Cap

Signed-off-by: Andreea Florescu <fandree@amazon.com>

andreeaflorescu committed:
renamed get_run_size to run_size

Signed-off-by: Andreea Florescu <fandree@amazon.com>

andreeaflorescu committed:
improved documentation on system ioctls

- Added examples for all ioctls.
- Added links to KVM documentation and internal links to KVM
  structures and functions for ease of access.
- Split documentation into summary and description.

Signed-off-by: Andreea Florescu <fandree@amazon.com>

andreeaflorescu committed:
added documentation for VM ioctls

- Added documentation for methods that take arguments.
- Added examples for public methods
- Added some details for ioctls where information was missing.

Signed-off-by: Andreea Florescu <fandree@amazon.com>

andreeaflorescu committed:
export NoDatamatch and IoEventAddress

These are used for calling VM ioctls from external crates.

Signed-off-by: Andreea Florescu <fandree@amazon.com>

alxiord reviewed (commented):
I ran `cargo doc --open` and have some general observations that don't necessarily pertain to these code changes but are documentation related nonetheless:

* The documentation for [`ioctl_ioc_nr`](https://github.com/rust-vmm/kvm-ioctls/blob/83c98b2724332c312cba7f79a2c1dbf3d089bf24/src/sys_ioctl.rs#L14) differs from other macros. It starts with `Raw macro to declare` whereas the others simply state `Declare...`. Please rephrase it.
```
ioctl_io_nr | Declare an ioctl that transfers no data.
ioctl_ioc_nr | Raw macro to declare a function that returns an ioctl number.
ioctl_ior_nr | Declare an ioctl that reads data.
ioctl_iow_nr | Declare an ioctl that writes data.
ioctl_iowr_nr | Declare an ioctl that reads and writes data.
```
* The documentation for wrappers states either `Wrapper for X` or `A wrapper for X`. I personally like the `Wrapper for` option better. In any case, please adjust the different ones.

```
CpuId | Wrapper for kvm_cpuid2 which has a zero length array at the end. Hides the zero length array behind a bounds check.
DeviceFd | Wrapper over the file descriptor obtained when creating an emulated device in the kernel.
Kvm | A wrapper over the /dev/kvm file.
KvmRunWrapper | A safe wrapper over the kvm_run struct.
VcpuFd | A wrapper around creating and using a kvm related vCPU file descriptor.
VmFd | A wrapper around creating and using a VM.
```

* The documentation for enums also differs: `CpuId` explicitly states that it is a list while `VcpuExit` states `Reasons`. Personal preference - I'd like `CpuId` to say `Capabilities exposed by KVM`. It's under the `Enums` category so it's bound to be a collection of something.

```
Cap | The list of capabilities exposed by KVM.
VcpuExit | Reasons for vCPU exits. The exit reasons are mapped to the KVM_EXIT_* defines from include/uapi/linux/kvm.h.
```

andreeaflorescu commented:
@aghecenco I will leave the ioctl declaration unchanged because having it inside the kvm-ioctls crate is a temporary state till vmm-sys-utils is merged. Once the crate vmm-sys-utils is merged we will remove the definition from this crate so I don't see a point in investing in the documentation just yet.

On the other points, I will take care of them once I get all the public interface documented. Right now, I only added documentation for cap.rs, system.rs and vm.rs is in progress. I'll ping you when I manage to document all items in the ioctls module so you can re-review it. Sorry if I haven't been upfront about it and wasted your time on reviewing. I will change the name of the PR to RFC.

andreeaflorescu committed:
enhanced the documentation of CpuId

- Added more usage examples
- Rephrased existing documentation

Signed-off-by: Andreea Florescu <fandree@amazon.com>

andreeaflorescu committed:
enhanced documentation of vCPU ioctls

- Added code examples
- Added references to the KVM API and KVM structures
- Added details for some ioctls

Signed-off-by: Andreea Florescu <fandree@amazon.com>

andreeaflorescu commented:
I think I am done with the documentation of the public interface. Please re-review @sameo @dianpopa @aghecenco
